taichi/backends/metal/kernel_util.h

```diff
@@ -63,48 +63,82 @@ struct KernelAttributes {
   std::string debug_string() const;
 };
 
+// This class contains the attributes descriptors for both the input args and
+// the return values of a Taichi kernel.
+//
 // Note that all Metal kernels belonging to the same Taichi kernel will share
-// the same kernel args (attributes + Metal buffer). This is because kernel
-// arguments is a Taichi-level concept.
+// the same kernel args (i.e. they use the same Metal buffer for input args and
+// return values). This is because kernel arguments is a Taichi-level concept.
+//
+// TODO(#909): Rename to KernelArgsRetsAttributes
 class KernelArgsAttributes {
- public:
-  // Attribute for a single argument.
-  // This is mostly the same as Kernel::Arg. It's extended to contain a few
-  // Metal kernel specific atrributes, like |stride| and |offset_in_mem|.
-  struct ArgAttributes {
+ private:
+  // Attributes that are shared by the input arg and the return value.
+  struct AttribsBase {
     // For array arg, this is #elements * stride(dt). Unit: byte
     size_t stride = 0;
     // Offset in the argument buffer
     size_t offset_in_mem = 0;
-    // Argument index
+    // Index of the input arg or the return value in the host `Context`
     int index = -1;
     MetalDataType dt;
     bool is_array = false;
-    bool is_return_val = false;
   };
 
-  explicit KernelArgsAttributes(const std::vector<Kernel::Arg> &args);
+ public:
+  // This is mostly the same as Kernel::Arg, with Metal specific attributes.
+  struct ArgAttributes : public AttribsBase {};
+
+  // This is mostly the same as Kernel::Ret, with Metal specific attributes.
+  struct RetAttributes : public AttribsBase {};
```

Comment:
```suggestion
  struct OutputAttributes : public AttribsBase {};
```

Reply:
FYI, I'd like to keep `Args` and `Ret`, and explained my opinion below. This way it is also consistent with `Kernel::Arg` and `Kernel::Ret`.